xgcm/test/test_grid.py

```diff
@@ -226,7 +226,10 @@ def _pad_right(data, boundary, fill_value=0.0):
     return np.hstack([data, pad_val])
 
 
-@pytest.mark.parametrize("boundary", [None, "extend", "fill"])
+@pytest.mark.parametrize(
+    "boundary",
+    [None, "extend", "fill", pytest.param("extrapolate", marks=pytest.mark.xfail)],
```

Comment:
Why should this fail? I am quite confused about this now.

Reply:
No idea. Seems to be for certain test datasets

```
=================================================== FAILURES ===================================================
________________ test_axis_neighbor_pairs_nonperiodic_1d[nonperiodic_1d_outer-True-extrapolate] ________________

nonperiodic_1d = (<xarray.Dataset>
Dimensions:  (XC: 9, XG: 10)
Coordinates:
  * XC       (XC) int64 1 2 3 4 5 6 7 8 9
  * XG       (XG...(XG) float64 0.3498 0.4589 0.009004 ... 0.08112 0.1661 0.4432, False, {'axes': {'X': {'center': 'XC', 'outer': 'XG'}}})
boundary = 'extrapolate', from_center = True

    @pytest.mark.parametrize(
        "boundary",
        [None, "extend", "fill", "extrapolate"],
    )
    @pytest.mark.parametrize("from_center", [True, False])
    def test_axis_neighbor_pairs_nonperiodic_1d(nonperiodic_1d, boundary, from_center):
        ds, periodic, expected = nonperiodic_1d
        axis = Axis(ds, "X", periodic=periodic)
    
        # detect whether this is an outer or inner case
        # outer --> dim_line_diff = 1
        # inner --> dim_line_diff = -1
        dim_len_diff = len(ds.XG) - len(ds.XC)
    
        if from_center:
            to = (set(expected["axes"]["X"].keys()) - {"center"}).pop()
            da = ds.data_c
        else:
            to = "center"
            da = ds.data_g
    
        shift = expected.get("shift") or False
    
        # need boundary condition for everything but outer to center
        if (boundary is None) and (
            dim_len_diff == 0
            or (dim_len_diff == 1 and from_center)
            or (dim_len_diff == -1 and not from_center)
        ):
            with pytest.raises(ValueError):
                data_left, data_right = axis._get_neighbor_data_pairs(
                    da, to, boundary=boundary
                )
        else:
            data_left, data_right = axis._get_neighbor_data_pairs(da, to, boundary=boundary)
            if ((dim_len_diff == 1) and not from_center) or (
                (dim_len_diff == -1) and from_center
            ):
                expected_left = da.data[:-1]
                expected_right = da.data[1:]
            elif ((dim_len_diff == 1) and from_center) or (
                (dim_len_diff == -1) and not from_center
            ):
                expected_left = _pad_left(da.data, boundary)
                expected_right = _pad_right(da.data, boundary)
            elif (shift and not from_center) or (not shift and from_center):
                expected_right = da.data
                expected_left = _pad_left(da.data, boundary)[:-1]
            else:
                expected_left = da.data
                expected_right = _pad_right(da.data, boundary)[1:]
    
>           np.testing.assert_allclose(data_left, expected_left)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=0
E           
E           Mismatched elements: 1 / 10 (10%)
E           Max absolute difference: 0.80105755
E           Max relative difference: 0.
E            x: array([0.801058, 0.860159, 0.919261, 0.83956 , 0.946847, 0.434604,
E                  0.027733, 0.812687, 0.147734, 0.875868])
E            y: array([0.      , 0.860159, 0.919261, 0.83956 , 0.946847, 0.434604,
E                  0.027733, 0.812687, 0.147734, 0.875868])

xgcm/test/test_grid.py:281: AssertionError
_______________ test_axis_neighbor_pairs_nonperiodic_1d[nonperiodic_1d_inner-False-extrapolate] ________________

nonperiodic_1d = (<xarray.Dataset>
Dimensions:  (XC: 9, XG: 8)
Coordinates:
  * XC       (XC) int64 1 2 3 4 5 6 7 8 9
  * XG       (XG)... float64 0.8048 0.0167 0.5203 0.6821 ... 0.2491 0.5931 0.3437, False, {'axes': {'X': {'center': 'XC', 'inner': 'XG'}}})
boundary = 'extrapolate', from_center = False

    @pytest.mark.parametrize(
        "boundary",
        [None, "extend", "fill", "extrapolate"],
    )
    @pytest.mark.parametrize("from_center", [True, False])
    def test_axis_neighbor_pairs_nonperiodic_1d(nonperiodic_1d, boundary, from_center):
        ds, periodic, expected = nonperiodic_1d
        axis = Axis(ds, "X", periodic=periodic)
    
        # detect whether this is an outer or inner case
        # outer --> dim_line_diff = 1
        # inner --> dim_line_diff = -1
        dim_len_diff = len(ds.XG) - len(ds.XC)
    
        if from_center:
            to = (set(expected["axes"]["X"].keys()) - {"center"}).pop()
            da = ds.data_c
        else:
            to = "center"
            da = ds.data_g
    
        shift = expected.get("shift") or False
    
        # need boundary condition for everything but outer to center
        if (boundary is None) and (
            dim_len_diff == 0
            or (dim_len_diff == 1 and from_center)
            or (dim_len_diff == -1 and not from_center)
        ):
            with pytest.raises(ValueError):
                data_left, data_right = axis._get_neighbor_data_pairs(
                    da, to, boundary=boundary
                )
        else:
            data_left, data_right = axis._get_neighbor_data_pairs(da, to, boundary=boundary)
            if ((dim_len_diff == 1) and not from_center) or (
                (dim_len_diff == -1) and from_center
            ):
                expected_left = da.data[:-1]
                expected_right = da.data[1:]
            elif ((dim_len_diff == 1) and from_center) or (
                (dim_len_diff == -1) and not from_center
            ):
                expected_left = _pad_left(da.data, boundary)
                expected_right = _pad_right(da.data, boundary)
            elif (shift and not from_center) or (not shift and from_center):
                expected_right = da.data
                expected_left = _pad_left(da.data, boundary)[:-1]
            else:
                expected_left = da.data
                expected_right = _pad_right(da.data, boundary)[1:]
    
>           np.testing.assert_allclose(data_left, expected_left)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=0
E           
E           Mismatched elements: 1 / 9 (11.1%)
E           Max absolute difference: 1.59283583
E           Max relative difference: 0.
E            x: array([1.592836, 0.804767, 0.016699, 0.520294, 0.682108, 0.92227 ,
E                  0.249147, 0.593147, 0.343706])
E            y: array([0.      , 0.804767, 0.016699, 0.520294, 0.682108, 0.92227 ,
E                  0.249147, 0.593147, 0.343706])

xgcm/test/test_grid.py:281: AssertionError
________________ test_axis_neighbor_pairs_nonperiodic_1d[nonperiodic_1d_left-True-extrapolate] _________________

nonperiodic_1d = (<xarray.Dataset>
Dimensions:  (XC: 100, XG: 100)
Coordinates:
  * XG       (XG) float64 0.0 0.06283 0.1257 0.1885 ......C) float64 0.207 0.05399 0.3635 0.4003 ... 0.573 0.3368 0.5606, False, {'axes': {'X': {'center': 'XC', 'left': 'XG'}}})
boundary = 'extrapolate', from_center = True

    @pytest.mark.parametrize(
        "boundary",
        [None, "extend", "fill", "extrapolate"],
    )
    @pytest.mark.parametrize("from_center", [True, False])
    def test_axis_neighbor_pairs_nonperiodic_1d(nonperiodic_1d, boundary, from_center):
        ds, periodic, expected = nonperiodic_1d
        axis = Axis(ds, "X", periodic=periodic)
    
        # detect whether this is an outer or inner case
        # outer --> dim_line_diff = 1
        # inner --> dim_line_diff = -1
        dim_len_diff = len(ds.XG) - len(ds.XC)
    
        if from_center:
            to = (set(expected["axes"]["X"].keys()) - {"center"}).pop()
            da = ds.data_c
        else:
            to = "center"
            da = ds.data_g
    
        shift = expected.get("shift") or False
    
        # need boundary condition for everything but outer to center
        if (boundary is None) and (
            dim_len_diff == 0
            or (dim_len_diff == 1 and from_center)
            or (dim_len_diff == -1 and not from_center)
        ):
            with pytest.raises(ValueError):
                data_left, data_right = axis._get_neighbor_data_pairs(
                    da, to, boundary=boundary
                )
        else:
            data_left, data_right = axis._get_neighbor_data_pairs(da, to, boundary=boundary)
            if ((dim_len_diff == 1) and not from_center) or (
                (dim_len_diff == -1) and from_center
            ):
                expected_left = da.data[:-1]
                expected_right = da.data[1:]
            elif ((dim_len_diff == 1) and from_center) or (
                (dim_len_diff == -1) and not from_center
            ):
                expected_left = _pad_left(da.data, boundary)
                expected_right = _pad_right(da.data, boundary)
            elif (shift and not from_center) or (not shift and from_center):
                expected_right = da.data
                expected_left = _pad_left(da.data, boundary)[:-1]
            else:
                expected_left = da.data
                expected_right = _pad_right(da.data, boundary)[1:]
    
>           np.testing.assert_allclose(data_left, expected_left)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=0
E           
E           Mismatched elements: 1 / 100 (1%)
E           Max absolute difference: 0.36003322
E           Max relative difference: 0.
E            x: array([0.360033, 0.207012, 0.053991, 0.363451, 0.400348, 0.597904,
E                  0.002926, 0.214173, 0.474725, 0.290928, 0.604397, 0.33844 ,
E                  0.981935, 0.95103 , 0.505543, 0.072431, 0.857241, 0.406657,...
E            y: array([0.      , 0.207012, 0.053991, 0.363451, 0.400348, 0.597904,
E                  0.002926, 0.214173, 0.474725, 0.290928, 0.604397, 0.33844 ,
E                  0.981935, 0.95103 , 0.505543, 0.072431, 0.857241, 0.406657,...

xgcm/test/test_grid.py:281: AssertionError
________________ test_axis_neighbor_pairs_nonperiodic_1d[nonperiodic_1d_left-False-extrapolate] ________________

nonperiodic_1d = (<xarray.Dataset>
Dimensions:  (XC: 100, XG: 100)
Coordinates:
  * XG       (XG) float64 0.0 0.06283 0.1257 0.1885 ......C) float64 0.207 0.05399 0.3635 0.4003 ... 0.573 0.3368 0.5606, False, {'axes': {'X': {'center': 'XC', 'left': 'XG'}}})
boundary = 'extrapolate', from_center = False

    @pytest.mark.parametrize(
        "boundary",
        [None, "extend", "fill", "extrapolate"],
    )
    @pytest.mark.parametrize("from_center", [True, False])
    def test_axis_neighbor_pairs_nonperiodic_1d(nonperiodic_1d, boundary, from_center):
        ds, periodic, expected = nonperiodic_1d
        axis = Axis(ds, "X", periodic=periodic)
    
        # detect whether this is an outer or inner case
        # outer --> dim_line_diff = 1
        # inner --> dim_line_diff = -1
        dim_len_diff = len(ds.XG) - len(ds.XC)
    
        if from_center:
            to = (set(expected["axes"]["X"].keys()) - {"center"}).pop()
            da = ds.data_c
        else:
            to = "center"
            da = ds.data_g
    
        shift = expected.get("shift") or False
    
        # need boundary condition for everything but outer to center
        if (boundary is None) and (
            dim_len_diff == 0
            or (dim_len_diff == 1 and from_center)
            or (dim_len_diff == -1 and not from_center)
        ):
            with pytest.raises(ValueError):
                data_left, data_right = axis._get_neighbor_data_pairs(
                    da, to, boundary=boundary
                )
        else:
            data_left, data_right = axis._get_neighbor_data_pairs(da, to, boundary=boundary)
            if ((dim_len_diff == 1) and not from_center) or (
                (dim_len_diff == -1) and from_center
            ):
                expected_left = da.data[:-1]
                expected_right = da.data[1:]
            elif ((dim_len_diff == 1) and from_center) or (
                (dim_len_diff == -1) and not from_center
            ):
                expected_left = _pad_left(da.data, boundary)
                expected_right = _pad_right(da.data, boundary)
            elif (shift and not from_center) or (not shift and from_center):
                expected_right = da.data
                expected_left = _pad_left(da.data, boundary)[:-1]
            else:
                expected_left = da.data
                expected_right = _pad_right(da.data, boundary)[1:]
    
            np.testing.assert_allclose(data_left, expected_left)
>           np.testing.assert_allclose(data_right, expected_right)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=0
E           
E           Mismatched elements: 1 / 100 (1%)
E           Max absolute difference: 0.51576146
E           Max relative difference: 0.
E            x: array([0.621719, 0.328495, 0.152055, 0.50959 , 0.573672, 0.28498 ,
E                  0.467737, 0.658653, 0.425918, 0.825677, 0.780014, 0.110298,
E                  0.689434, 0.57813 , 0.630895, 0.910186, 0.800797, 0.614423,...
E            y: array([0.621719, 0.328495, 0.152055, 0.50959 , 0.573672, 0.28498 ,
E                  0.467737, 0.658653, 0.425918, 0.825677, 0.780014, 0.110298,
E                  0.689434, 0.57813 , 0.630895, 0.910186, 0.800797, 0.614423,...

xgcm/test/test_grid.py:282: AssertionError
________________ test_axis_neighbor_pairs_nonperiodic_1d[nonperiodic_1d_right-True-extrapolate] ________________

nonperiodic_1d = (<xarray.Dataset>
Dimensions:  (XC: 100, XG: 100)
Coordinates:
  * XG       (XG) float64 0.06283 0.1257 0.1885 0.2513 ...3 0.8878 0.3353 0.2045 ... 0.6064 0.656 0.7242, False, {'axes': {'X': {'center': 'XC', 'right': 'XG'}}, 'shift': True})
boundary = 'extrapolate', from_center = True

    @pytest.mark.parametrize(
        "boundary",
        [None, "extend", "fill", "extrapolate"],
    )
    @pytest.mark.parametrize("from_center", [True, False])
    def test_axis_neighbor_pairs_nonperiodic_1d(nonperiodic_1d, boundary, from_center):
        ds, periodic, expected = nonperiodic_1d
        axis = Axis(ds, "X", periodic=periodic)
    
        # detect whether this is an outer or inner case
        # outer --> dim_line_diff = 1
        # inner --> dim_line_diff = -1
        dim_len_diff = len(ds.XG) - len(ds.XC)
    
        if from_center:
            to = (set(expected["axes"]["X"].keys()) - {"center"}).pop()
            da = ds.data_c
        else:
            to = "center"
            da = ds.data_g
    
        shift = expected.get("shift") or False
    
        # need boundary condition for everything but outer to center
        if (boundary is None) and (
            dim_len_diff == 0
            or (dim_len_diff == 1 and from_center)
            or (dim_len_diff == -1 and not from_center)
        ):
            with pytest.raises(ValueError):
                data_left, data_right = axis._get_neighbor_data_pairs(
                    da, to, boundary=boundary
                )
        else:
            data_left, data_right = axis._get_neighbor_data_pairs(da, to, boundary=boundary)
            if ((dim_len_diff == 1) and not from_center) or (
                (dim_len_diff == -1) and from_center
            ):
                expected_left = da.data[:-1]
                expected_right = da.data[1:]
            elif ((dim_len_diff == 1) and from_center) or (
                (dim_len_diff == -1) and not from_center
            ):
                expected_left = _pad_left(da.data, boundary)
                expected_right = _pad_right(da.data, boundary)
            elif (shift and not from_center) or (not shift and from_center):
                expected_right = da.data
                expected_left = _pad_left(da.data, boundary)[:-1]
            else:
                expected_left = da.data
                expected_right = _pad_right(da.data, boundary)[1:]
    
            np.testing.assert_allclose(data_left, expected_left)
>           np.testing.assert_allclose(data_right, expected_right)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=0
E           
E           Mismatched elements: 1 / 100 (1%)
E           Max absolute difference: 0.79234961
E           Max relative difference: 0.
E            x: array([0.88776 , 0.335336, 0.204521, 0.523932, 0.764641, 0.505969,
E                  0.575785, 0.16641 , 0.710131, 0.920735, 0.265737, 0.963353,
E                  0.042295, 0.734041, 0.587702, 0.46241 , 0.570693, 0.699676,...
E            y: array([0.88776 , 0.335336, 0.204521, 0.523932, 0.764641, 0.505969,
E                  0.575785, 0.16641 , 0.710131, 0.920735, 0.265737, 0.963353,
E                  0.042295, 0.734041, 0.587702, 0.46241 , 0.570693, 0.699676,...

xgcm/test/test_grid.py:282: AssertionError
_______________ test_axis_neighbor_pairs_nonperiodic_1d[nonperiodic_1d_right-False-extrapolate] ________________

nonperiodic_1d = (<xarray.Dataset>
Dimensions:  (XC: 100, XG: 100)
Coordinates:
  * XG       (XG) float64 0.06283 0.1257 0.1885 0.2513 ...3 0.8878 0.3353 0.2045 ... 0.6064 0.656 0.7242, False, {'axes': {'X': {'center': 'XC', 'right': 'XG'}}, 'shift': True})
boundary = 'extrapolate', from_center = False

    @pytest.mark.parametrize(
        "boundary",
        [None, "extend", "fill", "extrapolate"],
    )
    @pytest.mark.parametrize("from_center", [True, False])
    def test_axis_neighbor_pairs_nonperiodic_1d(nonperiodic_1d, boundary, from_center):
        ds, periodic, expected = nonperiodic_1d
        axis = Axis(ds, "X", periodic=periodic)
    
        # detect whether this is an outer or inner case
        # outer --> dim_line_diff = 1
        # inner --> dim_line_diff = -1
        dim_len_diff = len(ds.XG) - len(ds.XC)
    
        if from_center:
            to = (set(expected["axes"]["X"].keys()) - {"center"}).pop()
            da = ds.data_c
        else:
            to = "center"
            da = ds.data_g
    
        shift = expected.get("shift") or False
    
        # need boundary condition for everything but outer to center
        if (boundary is None) and (
            dim_len_diff == 0
            or (dim_len_diff == 1 and from_center)
            or (dim_len_diff == -1 and not from_center)
        ):
            with pytest.raises(ValueError):
                data_left, data_right = axis._get_neighbor_data_pairs(
                    da, to, boundary=boundary
                )
        else:
            data_left, data_right = axis._get_neighbor_data_pairs(da, to, boundary=boundary)
            if ((dim_len_diff == 1) and not from_center) or (
                (dim_len_diff == -1) and from_center
            ):
                expected_left = da.data[:-1]
                expected_right = da.data[1:]
            elif ((dim_len_diff == 1) and from_center) or (
                (dim_len_diff == -1) and not from_center
            ):
                expected_left = _pad_left(da.data, boundary)
                expected_right = _pad_right(da.data, boundary)
            elif (shift and not from_center) or (not shift and from_center):
                expected_right = da.data
                expected_left = _pad_left(da.data, boundary)[:-1]
            else:
                expected_left = da.data
                expected_right = _pad_right(da.data, boundary)[1:]
    
>           np.testing.assert_allclose(data_left, expected_left)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=0
E           
E           Mismatched elements: 1 / 100 (1%)
E           Max absolute difference: 0.87101381
E           Max relative difference: 0.
E            x: array([0.871014, 0.768644, 0.666275, 0.769796, 0.863174, 0.820112,
E                  0.100209, 0.274012, 0.648662, 0.335626, 0.125653, 0.49895 ,
E                  0.098243, 0.956442, 0.739064, 0.104403, 0.092836, 0.663784,...
E            y: array([0.      , 0.768644, 0.666275, 0.769796, 0.863174, 0.820112,
E                  0.100209, 0.274012, 0.648662, 0.335626, 0.125653, 0.49895 ,
E                  0.098243, 0.956442, 0.739064, 0.104403, 0.092836, 0.663784,...

xgcm/test/test_grid.py:281: AssertionError
========================================= 6 failed, 26 passed in 1.04s =========================================
```

Reply:
Ill have to dig deeper into this, but I am afraid my bandwidth is limited until the end of the week. Would it be ok for you to wait a few days? Sorry for this delay.

Reply:
Still not quite sure what is going on. But perhaps we should open a separate issue for getting to the bottom of 'extrapolate'. What do you think @rabernat?

Reply:
It's never been clear to me whether extrapolate was well supported or not. I'm fine with moving ahead on this PR as long as we have a separate issue to track the problems with extrapolate.